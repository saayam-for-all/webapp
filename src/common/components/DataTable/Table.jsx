import React, { useEffect, useMemo } from "react";
import { Link, useNavigate } from "react-router-dom";
import Pagination from "../Pagination/Pagination";

const Table = ({
  headers,
  rows,
  currentPage,
  setCurrentPage,
  totalPages,
  totalRows,
  itemsPerPage,
  sortConfig,
  requestSort,
  onRowsPerPageChange,
  getLinkPath,
  getLinkState = undefined,
}) => {
  const navigate = useNavigate();
  const paginatedRequests = useMemo(() => {
    return rows.slice(
      (currentPage - 1) * itemsPerPage,
      currentPage * itemsPerPage,
    );
  }, [rows, currentPage, itemsPerPage]);

  useEffect(() => {
    setCurrentPage(1);
  }, [totalRows, itemsPerPage]);

  const getSortIndicator = (key) => {
    if (sortConfig.key === key) {
      return sortConfig.direction === "ascending" ? "↑" : "↓";
    }
    return "";
  };

  return (
    <div className="relative h-full" data-testid="container">
      <div className="overflow-auto h-4/5">
        <table
          className="min-w-full divide-y divide-gray-200"
          data-testid="table"
        >
          <thead className="" data-testid="table-header">
            <tr>
              {headers.map((key) => (
                <th
                  key={key}
                  className="px-6 py-3 text-left text-xs font-medium text-gray-400 uppercase tracking-wider"
                  data-testid="map-header-one"
                >
                  <button type="button" onClick={() => requestSort(key)}>
                    {key.charAt(0).toUpperCase() +
                      key
                        .slice(1)
                        .replace(/([A-Z])/g, " $1")
                        .trim()}
                    {getSortIndicator(key)}
                  </button>
                </th>
              ))}
            </tr>
          </thead>
          <tbody
            className="bg-white divide-y divide-gray-200 "
            data-testid="table-body"
          >
            {paginatedRequests.map((request, rowIndex) => (
              <tr key={rowIndex}>
                {headers.map((header, colIndex) => (
                  <td
                    key={colIndex}
                    className="px-6 py-4 whitespace-nowrap"
                    data-testid="map-data-one"
                  >
                    {header === "id" ? (
                      <Link
                        to={getLinkPath(request, header)}
                        className="text-indigo-600 hover:text-indigo-900"
                        state={getLinkState ? getLinkState(request) : {}}
                      >
                        {request[header]}
                      </Link>
                    ) : (
                      request[header]
                    )}
                  </td>
                ))}
              </tr>
            ))}
          </tbody>
        </table>
      </div>
      <div className="h-1/5">
        {rows.length > 0 && (
          <Pagination
            currentPage={currentPage}
            totalPages={totalPages}
            onPageChange={setCurrentPage}
            rowsPerPage={itemsPerPage}
            totalRows={totalRows}
            onRowsPerPageChange={onRowsPerPageChange}
          />
        )}
      </div>
    </div>
  );
};

export default Table;
