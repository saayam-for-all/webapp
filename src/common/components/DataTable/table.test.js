import { render } from "@testing-library/react";
import Table from "./Table";

jest.mock("../Pagination/Pagination");

describe("Table", () => {
  const mockHeaders = ["id", "name"];
  const mockRows = [
    { id: 1, name: "Item 1" },
    { id: 2, name: "Item 2" },
  ];

  const mockSetCurrentPage = jest.fn();
  const mockRequestSort = jest.fn();
  const mockGetLinkPath = jest.fn(() => "/mock-link-path");
  const mockOnRowsPerPageChange = jest.fn();

  it("renders correctly", () => {
    const tree = render(
      <Table
        headers={mockHeaders}
        rows={mockRows}
        currentPage={1}
        setCurrentPage={mockSetCurrentPage}
        totalPages={2}
        totalRows={mockRows.length}
        itemsPerPage={2}
        sortConfig={{ key: "name", direction: "ascending" }}
        requestSort={mockRequestSort}
        onRowsPerPageChange={mockOnRowsPerPageChange}
        getLinkPath={mockGetLinkPath}
      />,
    );
    expect(tree).toMatchSnapshot();
  });
});
