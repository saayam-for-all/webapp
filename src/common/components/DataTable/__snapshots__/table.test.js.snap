// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="relative h-full"
        data-testid="container"
      >
        <div
          class="overflow-auto h-4/5"
        >
          <table
            class="min-w-full divide-y divide-gray-200"
            data-testid="table"
          >
            <thead
              class=""
              data-testid="table-header"
            >
              <tr>
                <th
                  class="px-6 py-3 text-left text-xs font-medium text-gray-400 uppercase tracking-wider"
                  data-testid="map-header-one"
                >
                  <button
                    type="button"
                  >
                    Id
                  </button>
                </th>
                <th
                  class="px-6 py-3 text-left text-xs font-medium text-gray-400 uppercase tracking-wider"
                  data-testid="map-header-one"
                >
                  <button
                    type="button"
                  >
                    Name
                    ↑
                  </button>
                </th>
              </tr>
            </thead>
            <tbody
              class="bg-white divide-y divide-gray-200 "
              data-testid="table-body"
            >
              <tr>
                <td
                  class="px-6 py-4 whitespace-nowrap"
                  data-testid="map-data-one"
                >
                  <mock-link
                    props="{"className":"text-indigo-600 hover:text-indigo-900","state":{}}"
                    to="/mock-link-path"
                  >
                    1
                  </mock-link>
                </td>
                <td
                  class="px-6 py-4 whitespace-nowrap"
                  data-testid="map-data-one"
                >
                  Item 1
                </td>
              </tr>
              <tr>
                <td
                  class="px-6 py-4 whitespace-nowrap"
                  data-testid="map-data-one"
                >
                  <mock-link
                    props="{"className":"text-indigo-600 hover:text-indigo-900","state":{}}"
                    to="/mock-link-path"
                  >
                    2
                  </mock-link>
                </td>
                <td
                  class="px-6 py-4 whitespace-nowrap"
                  data-testid="map-data-one"
                >
                  Item 2
                </td>
              </tr>
            </tbody>
          </table>
        </div>
        <div
          class="h-1/5"
        >
          <mock-pagination
            currentpage="1"
            rowsperpage="2"
            totalpages="2"
            totalrows="2"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="relative h-full"
      data-testid="container"
    >
      <div
        class="overflow-auto h-4/5"
      >
        <table
          class="min-w-full divide-y divide-gray-200"
          data-testid="table"
        >
          <thead
            class=""
            data-testid="table-header"
          >
            <tr>
              <th
                class="px-6 py-3 text-left text-xs font-medium text-gray-400 uppercase tracking-wider"
                data-testid="map-header-one"
              >
                <button
                  type="button"
                >
                  Id
                </button>
              </th>
              <th
                class="px-6 py-3 text-left text-xs font-medium text-gray-400 uppercase tracking-wider"
                data-testid="map-header-one"
              >
                <button
                  type="button"
                >
                  Name
                  ↑
                </button>
              </th>
            </tr>
          </thead>
          <tbody
            class="bg-white divide-y divide-gray-200 "
            data-testid="table-body"
          >
            <tr>
              <td
                class="px-6 py-4 whitespace-nowrap"
                data-testid="map-data-one"
              >
                <mock-link
                  props="{"className":"text-indigo-600 hover:text-indigo-900","state":{}}"
                  to="/mock-link-path"
                >
                  1
                </mock-link>
              </td>
              <td
                class="px-6 py-4 whitespace-nowrap"
                data-testid="map-data-one"
              >
                Item 1
              </td>
            </tr>
            <tr>
              <td
                class="px-6 py-4 whitespace-nowrap"
                data-testid="map-data-one"
              >
                <mock-link
                  props="{"className":"text-indigo-600 hover:text-indigo-900","state":{}}"
                  to="/mock-link-path"
                >
                  2
                </mock-link>
              </td>
              <td
                class="px-6 py-4 whitespace-nowrap"
                data-testid="map-data-one"
              >
                Item 2
              </td>
            </tr>
          </tbody>
        </table>
      </div>
      <div
        class="h-1/5"
      >
        <mock-pagination
          currentpage="1"
          rowsperpage="2"
          totalpages="2"
          totalrows="2"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
